src/state/VideoReducer/Video.reducer.jsx

```diff
@@ -21,14 +36,84 @@ function setVideos(state, payload) {
   };
 }
 
+function loadUserSettings(state) {
+  // console.log('[actions] loadUserSettings');
+
+  let savedSettings = storage.get(constants.REACT_CHALLENGE_USER_PREFERENCES);
+
+  if (savedSettings === null) {
+    savedSettings = { theme: 'light', favoriteVideos: { items: [] } };
+  }
+
+  // console.log('userSettings loaded', savedSettings);
+  return {
+    ...state,
+    favoriteVideos: savedSettings.favoriteVideos,
+    theme: savedSettings.theme,
+  };
+}
+
+function writeUserSettings(state) {
+  // console.log('[actions] writeUserSettings');
+  const settings = { ...state };
+  delete settings.query;
+  delete settings.videos;
+  delete settings.selectedVideo;
+
+  storage.set(constants.REACT_CHALLENGE_USER_PREFERENCES, settings);
+  // console.log('Settings saved', state);
+}
+
+const toggleFavoriteVideo = (state, video, isFavorite) => {
+  const newState = { ...state };
+  const favoriteIndex = isFavoriteVideo(video, newState.favoriteVideos);
+
+  if (!isFavorite) {
+    if (favoriteIndex === -1) {
+      const newVideo = {
+        ...video,
+        isFavorite: true,
+      };
+      newState.favoriteVideos.items.push(newVideo);
+      newState.selectedVideo = newVideo;
+    }
+  } else if (favoriteIndex > -1) {
+    const newVideo = {
+      ...video,
+      isFavorite: false,
+    };
+    newState.favoriteVideos.items.splice(favoriteIndex, 1);
+    newState.selectedVideo = newVideo;
+  }
+
+  // console.log(newState);
+  writeUserSettings(newState);
```

Comment:
for this part with the approach I propose to you, you can add an `useEffect` that `writesUserSettings` when the state changes since it's a side effect.

Reply:
It makes totally sense, thanks